src/Features/CSharp/Portable/Completion/CompletionProviders/SpeculativeTCompletionProvider.cs

```diff
@@ -58,46 +58,107 @@ private async Task<bool> ShouldShowSpeculativeTCompletionItemAsync(Document docu
+            var token = syntaxTree.FindTokenOnLeftOfPosition(position, cancellationToken);
+            var semanticModel = await document.GetSemanticModelForNodeAsync(token.Parent, cancellationToken).ConfigureAwait(false);
+
+            var spanStart = position;
+            while (true)
+            {
+                var orgSpanStart = spanStart;
+
+                spanStart = WalkOutOfGenericType(syntaxTree, spanStart, semanticModel, cancellationToken);
+                spanStart = WalkOutOfTupleType(syntaxTree, spanStart, cancellationToken);
+                spanStart = WalkOutOfRefType(syntaxTree, spanStart, cancellationToken);
+
+                if (spanStart == orgSpanStart)
+                {
+                    break;
+                }
+            }
+
+            if (IsStartOfSpeculativeTContext(syntaxTree, spanStart, allowAsyncKeyword: spanStart != position, cancellationToken))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        private static bool IsStartOfSpeculativeTContext(SyntaxTree syntaxTree, int position, bool allowAsyncKeyword, CancellationToken cancellationToken)
+        {
+            var token = syntaxTree.FindTokenOnLeftOfPosition(position, cancellationToken);
+            var prevToken = token.GetPreviousTokenIfTouchingWord(position);
+
+            if (((allowAsyncKeyword || !prevToken.IsKindOrHasMatchingText(SyntaxKind.AsyncKeyword)) &&
+                syntaxTree.IsMemberDeclarationContext(position, contextOpt: null, validModifiers: SyntaxKindSet.AllMemberModifiers, validTypeDeclarations: SyntaxKindSet.ClassInterfaceStructTypeDeclarations, canBePartial: true, cancellationToken: cancellationToken)) ||
+                syntaxTree.IsStatementContext(position, token, cancellationToken) ||
+                syntaxTree.IsGlobalMemberDeclarationContext(position, SyntaxKindSet.AllGlobalMemberModifiers, cancellationToken) ||
+                syntaxTree.IsGlobalStatementContext(position, cancellationToken) ||
+                syntaxTree.IsDelegateReturnTypeContext(position, token, cancellationToken))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        private static int WalkOutOfGenericType(SyntaxTree syntaxTree, int position, SemanticModel semanticModel, CancellationToken cancellationToken)
+        {
+            var spanStart = position;
+            var token = syntaxTree.FindTokenOnLeftOfPosition(position, cancellationToken);
+
+            if (syntaxTree.IsGenericTypeArgumentContext(position, token, cancellationToken, semanticModel))
             {
                 // Walk out until we find the start of the partial written generic
-                while (syntaxTree.IsInPartiallyWrittenGeneric(testPosition, cancellationToken, out var nameToken))
+                while (syntaxTree.IsInPartiallyWrittenGeneric(spanStart, cancellationToken, out var nameToken))
```

Comment:
Do we need the loop in this case? Or since you already have an outer loop that's calling WalkOutOfGenericType is that good enough?

Reply:
👍

Reply:
I don't care, changing to `if`.